File: backend/src/main/java/com/improvementsolutions/service/Iso9001CatalogService.java
package com.improvementsolutions.service;

import com.improvementsolutions.model.Iso9001CatalogItem;
import com.improvementsolutions.repository.Iso9001CatalogItemRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.regex.Pattern;

@Service
@RequiredArgsConstructor
public class Iso9001CatalogService {

    private static final Set<String> ALLOWED_CATALOG_CODES = Set.of(
            "tipo-documento",
            "proceso",
            "codigo",
            "almacenamiento",
            "disposicion-final"
    );

    /** Catálogos donde el código corto es obligatorio (número de registro SGC). */
    private static final Set<String> CODE_REQUIRED_CATALOGS = Set.of("tipo-documento", "proceso");

    private static final Pattern ITEM_CODE_PATTERN = Pattern.compile("^[A-Z]{2,5}$");

    private final Iso9001CatalogItemRepository repository;

    public void assertCatalogCode(String catalogCode) {
        if (catalogCode == null || !ALLOWED_CATALOG_CODES.contains(catalogCode.trim())) {
            throw new IllegalArgumentException("Código de catálogo ISO 9001 no válido: " + catalogCode);
        }
    }

    public List<Iso9001CatalogItem> findAllByCatalog(String catalogCode) {
        assertCatalogCode(catalogCode);
        return repository.findByCatalogCodeOrderByIdAsc(catalogCode.trim());
    }

    public Optional<Iso9001CatalogItem> findByCatalogAndId(String catalogCode, Long id) {
        assertCatalogCode(catalogCode);
        return repository.findByCatalogCodeAndId(catalogCode.trim(), id);
    }

    @Transactional
    public Iso9001CatalogItem create(String catalogCode, Iso9001CatalogItem input) {
        assertCatalogCode(catalogCode);
        String cc = catalogCode.trim();
        String name = normalizeName(input.getName());
        String code = normalizeItemCode(cc, input.getCode());
        repository.findByCatalogCodeAndName(cc, name).ifPresent(x -> {
            throw new IllegalStateException("DUPLICATE_NAME");
        });
        assertCodeUnique(cc, code, null);
        Iso9001CatalogItem entity = new Iso9001CatalogItem();
        entity.setCatalogCode(cc);
        entity.setCode(code);
        entity.setName(name);
        entity.setDescription(trimToNull(input.getDescription()));
        return repository.save(entity);
    }

    @Transactional
    public Iso9001CatalogItem update(String catalogCode, Long id, Iso9001CatalogItem input) {
        assertCatalogCode(catalogCode);
        String cc = catalogCode.trim();
        Iso9001CatalogItem existing = repository.findByCatalogCodeAndId(cc, id)
                .orElseThrow(() -> new IllegalStateException("NOT_FOUND"));
        String name = normalizeName(input.getName());
        String code = normalizeItemCode(cc, input.getCode());
        if (repository.existsByCatalogCodeAndNameAndIdNot(cc, name, id)) {
            throw new IllegalStateException("DUPLICATE_NAME");
        }
        assertCodeUnique(cc, code, id);
        existing.setName(name);
        existing.setCode(code);
        existing.setDescription(trimToNull(input.getDescription()));
        return repository.save(existing);
    }

    @Transactional
    public void delete(String catalogCode, Long id) {
        assertCatalogCode(catalogCode);
        Iso9001CatalogItem existing = repository.findByCatalogCodeAndId(catalogCode.trim(), id)
                .orElseThrow(() -> new IllegalStateException("NOT_FOUND"));
        repository.delete(existing);
    }

    private void assertCodeUnique(String catalogCode, String code, Long excludeId) {
        if (code == null) {
            return;
        }
        if (excludeId == null) {
            repository.findByCatalogCodeAndCodeIgnoreCase(catalogCode, code).ifPresent(x -> {
                throw new IllegalStateException("DUPLICATE_CODE");
            });
        } else if (repository.existsByCatalogCodeAndCodeIgnoreCaseAndIdNot(catalogCode, code, excludeId)) {
            throw new IllegalStateException("DUPLICATE_CODE");
        }
    }

    /**
     * Normaliza el código a mayúsculas (2–5 letras A–Z).
     * Obligatorio en tipo-documento; opcional en el resto.
     */
    private static String normalizeItemCode(String catalogCode, String raw) {
        if (raw == null || raw.isBlank()) {
            if (CODE_REQUIRED_CATALOGS.contains(catalogCode)) {
                throw new IllegalArgumentException("El código es obligatorio (2 a 5 letras, p. ej. PRO, MAN, FOR)");
            }
            return null;
        }
        String code = raw.trim().toUpperCase();
        if (!ITEM_CODE_PATTERN.matcher(code).matches()) {
            throw new IllegalArgumentException("El código debe tener entre 2 y 5 letras (A–Z), sin números ni espacios");
        }
        return code;
    }

    private static String normalizeName(String name) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("El nombre es obligatorio");
        }
        return name.trim();
    }

    private static String trimToNull(String description) {
        if (description == null || description.isBlank()) {
            return null;
        }
        return description.trim();
    }
}

File: frontend/src/app/services/iso-9001-catalog.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export type Iso9001CatalogKey =
  | 'tipo-documento'
  | 'proceso'
  | 'codigo'
  | 'almacenamiento'
  | 'disposicion-final';

/** Ítem del catálogo global ISO 9001 (configuración central). */
export interface Iso9001CatalogItem {
  id?: number;
  name: string;
  /** Código corto (2–5 letras) para el número de registro del SGC. */
  code?: string | null;
  description?: string;
  catalogCode?: string;
  createdAt?: string;
  updatedAt?: string;
}

@Injectable({
  providedIn: 'root'
})
export class Iso9001CatalogService {
  private readonly baseUrl = `${environment.apiUrl}/api/master-data/iso-9001-catalog`;

  constructor(private readonly http: HttpClient) {}

  getAll(catalogKey: Iso9001CatalogKey): Observable<Iso9001CatalogItem[]> {
    return this.http.get<Iso9001CatalogItem[]>(`${this.baseUrl}/${catalogKey}`);
  }

  getById(catalogKey: Iso9001CatalogKey, id: number): Observable<Iso9001CatalogItem> {
    return this.http.get<Iso9001CatalogItem>(`${this.baseUrl}/${catalogKey}/${id}`);
  }

  create(
    catalogKey: Iso9001CatalogKey,
    body: Pick<Iso9001CatalogItem, 'name' | 'description' | 'code'>
  ): Observable<Iso9001CatalogItem> {
    return this.http.post<Iso9001CatalogItem>(`${this.baseUrl}/${catalogKey}`, body);
  }

  update(
    catalogKey: Iso9001CatalogKey,
    id: number,
    body: Pick<Iso9001CatalogItem, 'name' | 'description' | 'code'>
  ): Observable<Iso9001CatalogItem> {
    return this.http.put<Iso9001CatalogItem>(`${this.baseUrl}/${catalogKey}/${id}`, body);
  }

  delete(catalogKey: Iso9001CatalogKey, id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${catalogKey}/${id}`);
  }
}
